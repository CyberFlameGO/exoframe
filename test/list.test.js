/* eslint-env jest */
// mock config for testing
jest.mock('../src/config', () => require('./__mocks__/config'));

// npm packages
const nock = require('nock');
const sinon = require('sinon');

// our packages
const {handler: list} = require('../src/commands/list');

const containers = [
  {
    Id: '123',
    Name: '/test',
    Config: {
      Labels: {
        'traefik.http.routers.test.rule': 'Host(`test.host`)',
        'exoframe.deployment': 'test',
        'exoframe.project': 'test',
      },
    },
    State: {
      Status: 'Up 10 minutes',
    },
    NetworkSettings: {
      Networks: {
        exoframe: {
          Aliases: null,
        },
      },
    },
  },
  {
    Id: '321',
    Name: '/test2',
    Config: {
      Labels: {'exoframe.project': 'test'},
    },
    State: {
      Status: 'Up 12 minutes',
    },
    NetworkSettings: {
      Networks: {
        exoframe: {
          Aliases: null,
        },
      },
    },
  },
  {
    Id: '111',
    Name: '/test3',
    Config: {
      Labels: {'exoframe.project': 'other'},
    },
    State: {
      Status: 'Up 13 minutes',
    },
    NetworkSettings: {
      Networks: {
        exoframe: {
          Aliases: null,
        },
      },
    },
  },
  {
    Id: '444',
    Name: '/test4',
    Config: {
      Labels: {'exoframe.project': 'somethingelse'},
    },
    State: {
      Status: 'Up 10 minutes',
    },
    NetworkSettings: {
      Networks: {
        default: {
          Aliases: null,
        },
        traefik: {
          Aliases: ['alias4'],
        },
      },
    },
  },
];

const services = [
  {
    ID: '12345',
    Spec: {
      Name: 'test-service-one',
      Labels: {
        'exoframe.project': 'test-service',
      },
      Networks: [
        {
          Target: 'netid',
        },
      ],
    },
  },
  {
    ID: '0987',
    Spec: {
      Name: 'test-service-two',
      Labels: {
        'exoframe.project': 'test-service',
        'exoframe.deployment': 'test-service-two',
        'traefik.http.routers.test-service-two.rule': 'Host(`test.host`)',
      },
      Networks: [
        {
          Target: 'netid',
          Aliases: ['test.host'],
        },
      ],
    },
  },
  {
    ID: '321',
    Spec: {
      Name: 'test-service-three',
      Labels: {
        'exoframe.project': 'test-project',
        'exoframe.deployment': 'test-service-three',
        'traefik.http.routers.test-service-three.rule': 'Host(`other.domain`)',
      },
      Networks: [
        {
          Target: 'netid',
        },
      ],
    },
  },
];

// test list
test('Should get list of deployments', done => {
  // handle correct request
  const listServer = nock('http://localhost:8080')
    .get(`/list`)
    .reply(200, {containers});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // execute login
  list().then(() => {
    // make sure log in was successful
    // check that server was called
    expect(listServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    listServer.done();
    done();
  });
});

// test swarm list
test('Should get list of swarm deployments', done => {
  // handle correct request
  const listServer = nock('http://localhost:8080')
    .get(`/list`)
    .reply(200, {services});
  // spy on console
  const consoleSpy = sinon.spy(console, 'log');
  // execute login
  list().then(() => {
    // make sure log in was successful
    // check that server was called
    expect(listServer.isDone()).toBeTruthy();
    // first check console output
    expect(consoleSpy.args).toMatchSnapshot();
    // restore console
    console.log.restore();
    listServer.done();
    done();
  });
});
