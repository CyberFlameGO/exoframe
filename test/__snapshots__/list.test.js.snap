// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should get list of deployments 1`] = `
Array [
  Array [
    "[32m4 deployments found on http://localhost:8080:[39m
[32m[39m",
  ],
  Array [
    "> [34m[1m[4mNormal[24m[22m[39m deployments:
",
  ],
  Array [
    "Deployments for [1mtest[22m:",
  ],
  Array [],
  Array [
    "[31m   ID      [39m[90m [39m[31m   URL         [39m[90m [39m[31m   Hostname   [39m[90m [39m[31m   Status          [39m[90m [39m[31m               [39m
   test    [90m [39m   test.host   [90m [39m   Not set    [90m [39m   Up 10 minutes   [90m [39m   Container   
   test2   [90m [39m   Not set     [90m [39m   Not set    [90m [39m   Up 12 minutes   [90m [39m   Container   ",
  ],
  Array [],
  Array [
    "Other deployments:",
  ],
  Array [],
  Array [
    "[31m   ID      [39m[90m [39m[31m   URL       [39m[90m [39m[31m   Hostname   [39m[90m [39m[31m   Status          [39m[90m [39m[31m   Type        [39m
   test3   [90m [39m   Not set   [90m [39m   Not set    [90m [39m   Up 13 minutes   [90m [39m   Container   
   test4   [90m [39m   Not set   [90m [39m   alias4     [90m [39m   Up 10 minutes   [90m [39m   Container   ",
  ],
]
`;

exports[`Should get list of swarm deployments 1`] = `
Array [
  Array [
    "[32m3 deployments found on http://localhost:8080:[39m
[32m[39m",
  ],
  Array [
    "> [34m[1m[4mSwarm mode[24m[22m[39m deployments:
",
  ],
  Array [
    "Deployments for [1mtest-service[22m:",
  ],
  Array [],
  Array [
    "[31m   ID                 [39m[90m [39m[31m   URL         [39m[90m [39m[31m   Hostname    [39m[90m [39m[31m   Status   [39m[90m [39m[31m      [39m
   test-service-one   [90m [39m   Not set     [90m [39m   Not set     [90m [39m            [90m [39m      
   test-service-two   [90m [39m   test.host   [90m [39m   test.host   [90m [39m            [90m [39m      ",
  ],
  Array [],
  Array [
    "Other deployments:",
  ],
  Array [],
  Array [
    "[31m   ID                   [39m[90m [39m[31m   URL            [39m[90m [39m[31m   Hostname   [39m[90m [39m[31m   Status   [39m[90m [39m[31m   Type   [39m
   test-service-three   [90m [39m   other.domain   [90m [39m   Not set    [90m [39m            [90m [39m          ",
  ],
]
`;
